apps/condo/domains/organization/sbbol/oauth2.js

```diff
@@ -0,0 +1,170 @@
+// Todo(zuch): need to write JWT verification
+
+const conf = process.env
+
+const { Issuer, custom } = require('openid-client') // certified openid client will all checks
+const jwtDecode = require('jwt-decode') // decode jwt without validation
+
+const SBBOL_CONFIG = conf.SBBOL_CONFIG ? JSON.parse(conf.SBBOL_CONFIG) : {}
+const SBBOL_PFX = conf.SBBOL_PFX ? JSON.parse(conf.SBBOL_PFX) : {}
+const JWT_ALG = 'gost34.10-2012'
+
+class SbbolOauth2Api {
+    constructor () {
+        this.host = SBBOL_CONFIG.host
+        this.protectedHost = SBBOL_CONFIG.protected_host
+        this.port = SBBOL_CONFIG.port
+        this.clientSecret = SBBOL_CONFIG.client_secret
+        this.clientId = SBBOL_CONFIG.client_id
+        this.serviceId = SBBOL_CONFIG.service_id
+        this.alg = JWT_ALG
+        this.redirectUrl = `${conf.SERVER_URL}/api/sbbol/auth/callback`
+        if (SBBOL_PFX.certificate) {
+            this.certificate = {
+                pfx: Buffer.from(SBBOL_PFX.certificate, 'base64'),
+                passphrase: SBBOL_PFX.passphrase,
+            }
+        }
+        this.createClient()
+    }
+
+    createClient () {
+        this.createIssuer()
+        const client = new this.issuer.Client({
+            client_id: this.clientId + '',
+            client_secret: this.clientSecret,
+            redirect_uris: [this.redirectUrl],
+            response_types: ['code'],
+            authorization_signed_response_alg: this.alg,
+            id_token_signed_response_alg: this.alg,
+            userinfo_signed_response_alg: this.alg,
+            token_endpoint_auth_method: 'client_secret_post',
+            tls_client_certificate_bound_access_tokens: true,
+        })
+        client[custom.http_options] = (options) => {
+            let withCertificate = { ...options }
+            if (this.certificate) {
+                const { pfx, passphrase } = this.certificate
+                withCertificate = { ...withCertificate, https: { pfx, passphrase } }
+            }
+            return withCertificate
+        }
+        // we override standart JWT validation as we do not have JWK from oauth provider
+        client.validateJWT = async (jwt) => {
+            const [ header, payload ] = [jwtDecode(jwt, { header: true }), jwtDecode(jwt)]
+            return { protected: header, payload }
+        }
+        this.client = client
+    }
+
+    createIssuer () {
+        const sbbolIssuer = new Issuer({
+            issuer: this.protectedUrl,
+            authorization_endpoint: this.authUrl,
+            token_endpoint: this.tokenUrl,
+            userinfo_endpoint: this.userInfoUrl,
+            revocation_endpoint: this.revokeUrl,
+        })
+        // turn off JWKS storage as it's not workin with sbbol
+        sbbolIssuer.keystore = async () => null
+        sbbolIssuer.queryKeyStore = async () => null
```

Comment:
can we move it to initialization?

Reply:
no - these methods are an override of a third party's libraries and are not present in the constructor